app/src/main/java/org/eyeseetea/malariacare/data/database/iomodules/dhis/importer/ConvertFromSDKVisitor.java

```diff
@@ -584,9 +584,12 @@ public void buildOrgUnitProgramRelation(OrganisationUnitExtended sdkOrganisation
         for (ProgramExtended sdkProgramExtended : ProgramExtended.getExtendedList(
                 SdkQueries.getProgramsForOrganisationUnit(appOrgUnit.getUid(),
                         ProgramType.WITHOUT_REGISTRATION))) {
-            sdkProgramExtended.setAppProgram(programMapObjects.get(sdkProgramExtended.getUid()));
+            if(sdkProgramExtended.isValidProgram()) {
+                sdkProgramExtended.setAppProgram(
+                        programMapObjects.get(sdkProgramExtended.getUid()));
 
-            addOrgUnitProgramRelation(sdkOrganisationUnitExtended, sdkProgramExtended);
+                addOrgUnitProgramRelation(sdkOrganisationUnitExtended, sdkProgramExtended);
+            }
```

Comment:
Can we loop here not on all the programs, but only on valid programs. I would change the getProgramsForOrganisationUnit by adding an optional parameter (function overloading) to determine the program type